File: typescript/src/chain-of-responsibility/chain-of-responsibility.test.ts
import { describe, expect, it } from "vitest";
import type { Handler } from "./handler.ts";
import { NegativeHandler } from "./negative-handler.ts";
import { PositiveHandler } from "./positive-handler.ts";
import { ZeroHandler } from "./zero-handler.ts";

function fullChain(): Handler {
  const head = new NegativeHandler();
  head.setNext(new ZeroHandler()).setNext(new PositiveHandler());
  return head;
}

describe("ChainOfResponsibility", () => {
  it("each request is answered by the responsible handler", () => {
    const chain = fullChain();
    expect(chain.handle(-5)).toBe("negative");
    expect(chain.handle(0)).toBe("zero");
    expect(chain.handle(7)).toBe("positive");
  });

  it("a request nobody claims falls off the end", () => {
    const shortened = new NegativeHandler();
    shortened.setNext(new ZeroHandler());
    expect(shortened.handle(1)).toBe("unhandled");
    expect(new PositiveHandler().handle(-1)).toBe("unhandled");
  });

  it("relinking changes who answers", () => {
    const head = new PositiveHandler();
    expect(head.handle(-1)).toBe("unhandled");
    head.setNext(new NegativeHandler());
    expect(head.handle(-1)).toBe("negative");
  });
});

File: java/src/test/java/com/penapereira/patterns/chainofresponsibility/ChainOfResponsibilityTest.java
package com.penapereira.patterns.chainofresponsibility;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class ChainOfResponsibilityTest {

    private static Handler fullChain() {
        Handler head = new NegativeHandler();
        head.setNext(new ZeroHandler()).setNext(new PositiveHandler());
        return head;
    }

    @Test
    void eachRequestIsAnsweredByTheResponsibleHandler() {
        Handler chain = fullChain();
        assertEquals("negative", chain.handle(-5));
        assertEquals("zero", chain.handle(0));
        assertEquals("positive", chain.handle(7));
    }

    @Test
    void aRequestNobodyClaimsFallsOffTheEnd() {
        Handler shortened = new NegativeHandler();
        shortened.setNext(new ZeroHandler());
        assertEquals("unhandled", shortened.handle(1));
        assertEquals("unhandled", new PositiveHandler().handle(-1));
    }

    @Test
    void relinkingChangesWhoAnswers() {
        Handler head = new PositiveHandler();
        assertEquals("unhandled", head.handle(-1));
        head.setNext(new NegativeHandler());
        assertEquals("negative", head.handle(-1));
    }
}
